crates/subspace-farmer/src/utils/segment_reconstruction.rs

```diff
@@ -0,0 +1,79 @@
+use std::iter;
+use futures::stream::{FuturesOrdered, FuturesUnordered};
+use futures::StreamExt;
+use subspace_core_primitives::{Piece, PieceIndex, SegmentIndex, PIECES_IN_SEGMENT, RECORD_SIZE};
+use subspace_farmer_components::plotting::PieceGetter;
+use tracing::{error, info};
+use subspace_archiving::reconstructor::Reconstructor;
+
+const SEGMENT_SIZE: u32 = RECORD_SIZE * PIECES_IN_SEGMENT / 2;
+
+pub struct SegmentReconstruction<PG> {
+    piece_getter: PG,
+}
+
+impl<PG: PieceGetter> SegmentReconstruction<PG> {
+    pub fn new(piece_getter: PG) -> Self {
+        Self { piece_getter }
+    }
+
+    pub async fn get_missing_piece(&self, missing_piece_index: PieceIndex) -> Option<Piece> {
+        let segment_index: SegmentIndex =
+            missing_piece_index / SegmentIndex::from(PIECES_IN_SEGMENT);
+
+        // TODO: Consider taking more pieces here.
+        // We take minimum pieces to reconstruct the segment plus one to mitigate the missing piece.
+        let pieces_to_retrieve = (PIECES_IN_SEGMENT / 2 + 1) as usize;
+
+        let starting_piece_index = segment_index * SegmentIndex::from(PIECES_IN_SEGMENT);
+
+        let piece_offset = (missing_piece_index - starting_piece_index) as usize; // TODO:
+        info!(%missing_piece_index, %starting_piece_index, %pieces_to_retrieve, %piece_offset);
+
+        let piece_indexes = (starting_piece_index..)
+            .take(pieces_to_retrieve)
+            .collect::<Vec<_>>();
```

Comment:
I'd take all the pieces in segment (except the one we're missing), shuffle them all and then retrieve in a batch, stopping once we've reached half of the segment pieces. Only trying to download 1/2+1 means this can still fail when network is perfectly capable of reconstructing the segment. I think we can use semaphores here to ensure we only download as many things in parallel as necessary. For instance once we've got 1/2-1 pieces we should only have 1 request running and the rest should be blocked until we are confident that the request we're running right now has failed.

The other way to describe this is that there will be 256 futures in the `piece_requests` below, but 1/2 of them will resolve with `None` (one of which is `missing_piece_index`). Then we call reconstructor and get all of them back.